edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/AuthAgentRequestHandler.cs

```diff
@@ -0,0 +1,355 @@
+                }
+                else if (isCertificatePresent)
+                {
+                    var certificate = DecodeCertificate(request.EncodedCertificate);
+                    var chain = DecodeCertificateChain(request.EncodedCertificateChain);
+
+                    result = Option.Some(this.clientCredentialsFactory.GetWithX509Cert(deviceId, moduleId, deviceClientType, certificate, chain));
+                }
+                else
+                {
+                    Events.NoCredentialsSpecified();
+                }
+            }
+            catch (Exception e)
+            {
+                Events.InvalidCredentials(e);
+            }
+
+            return result;
+        }
+
+        async Task<bool> AuthenticateAsync(IClientCredentials credentials)
+        {
+            try
+            {
+                return await this.authenticator.AuthenticateAsync(credentials);
+            }
+            catch (Exception e)
+            {
+                Events.InvalidCredentials(e);
+            }
+
+            return false;
+        }
+
+        bool IsBaseUrl(PathString path) => path.HasValue && string.Equals(path.Value, this.config.BaseUrl, StringComparison.Ordinal);
+
+        static Task SendAuthResultAsync(HttpContext context, bool isAuthenticated, Option<IClientCredentials> credentials)
+        {
+            // note, that if authenticated, then these values are present, and defaults never apply
+            var iotHubName = credentials.Map(c => c.Identity.IotHubHostName).GetOrElse("any");
+            var id = credentials.Map(c => c.Identity.Id).GetOrElse("anonymous");
+
+            if (isAuthenticated)
+            {
+                Events.AuthSucceeded(id);
+                return SendAsync(context, $@"{{""result"":200,""identity"":""{iotHubName}/{id}"",""version"":""2020-04-20""}}");
+            }
+            else
+            {
+                Events.AuthFailed(id);
+                return SendErrorAsync(context);
+            }
+        }
+
+        static Task SendErrorAsync(HttpContext context)
+        {
+            // we don't reveal error details: for the caller if something is not right, that is unauthenticated
+            return SendAsync(context, @"{""result"":403,""version"":""2020-04-20""}");
+        }
+
+        static async Task SendAsync(HttpContext context, string body)
+        {
+            byte[] buffer = Encoding.UTF8.GetBytes(body);
+
+            var mediaType = new MediaTypeHeaderValue("application/json");
+            mediaType.Encoding = Encoding.UTF8;
+            context.Response.ContentType = mediaType.ToString();
+            context.Response.ContentLength = buffer.Length;
+
+            await context.Response.Body.WriteAsync(buffer, 0, buffer.Length);
+        }
+
+        static async Task<Option<string>> ReadRequestBodyAsStringAsync(HttpRequest request)
```

Comment:
can you please clarify why you need all this machinery? Isn't there already a way in asp.net core to read body buffered?

Reply:
The story of this handler that first it was built around a http listener, but then it turned out that it has permission problems on windows (needs admin), so I turned the same code into kestrel based, but kept the manual work which was already implemented. I can turn it into an "official" webapi with controllers and routing, but I felt unnecessary

Reply:
changed it to controller based